{
  try {
    T res=fut.get();
    if (fut.isCancelled())     throw new CancellationException("Task was cancelled: " + fut);
    return res;
  }
 catch (  GridException e) {
    if (fut.isCancelled()) {
      RuntimeException ex=new CancellationException("Task was cancelled: " + fut);
      ex.initCause(e);
      throw ex;
    }
    throw new ExecutionException("Failed to get task result: " + fut,e);
  }
}
